File: examples/go/apps/agent/mcp_stdio.go
package agent

import (
	"bufio"
	"encoding/json"
	"os"

	"app-protocol/examples/go/core/shared"
)

type StdioAdapter struct{}

func NewStdioAdapter() *StdioAdapter {
	return &StdioAdapter{}
}

func (a *StdioAdapter) Transport() string {
	return "stdio"
}

func (a *StdioAdapter) Serve(server shared.AppMcpServer) error {
	sessionID := shared.NewID()
	phase := "awaiting_initialize"
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)

	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	for scanner.Scan() {
		line := scanner.Text()
		if stringsTrim(line) == "" {
			continue
		}

		var message any
		if err := json.Unmarshal([]byte(line), &message); err != nil {
			if err := writeJSONRPCResponse(writer, failureResponse(nil, -32700, "Invalid JSON-RPC payload.", err.Error())); err != nil {
				return err
			}
			continue
		}

		request, hasID, ok := parseJSONRPCMessage(message)
		if !ok {
			if err := writeJSONRPCResponse(writer, failureResponse(nil, -32600, "Invalid JSON-RPC request shape.", nil)); err != nil {
				return err
			}
			continue
		}

		if !hasID {
			if request.Method == "notifications/initialized" && phase == "awaiting_initialized_notification" {
				phase = "ready"
			}
			continue
		}

		context := &shared.AppMcpRequestContext{
			Transport: "stdio",
			SessionID: sessionID,
		}

		response := a.handleRequestWithPhase(request, server, context, &phase)
		if err := writeJSONRPCResponse(writer, response); err != nil {
			return err
		}
	}

	return scanner.Err()
}

func (a *StdioAdapter) handleRequestWithPhase(request jsonRPCRequest, server shared.AppMcpServer, context *shared.AppMcpRequestContext, phase *string) jsonRPCResponse {
	switch request.Method {
	case "initialize":
		if *phase != "awaiting_initialize" {
			return failureResponse(request.ID, -32600, "MCP initialize may only run once per stdio session.", nil)
		}
		params := &shared.AppMcpInitializeParams{}
		if request.Params != nil {
			encoded, _ := json.Marshal(request.Params)
			_ = json.Unmarshal(encoded, params)
		}
		result, err := server.Initialize(params, context)
		if err != nil {
			return responseFromError(request.ID, err)
		}
		*phase = "awaiting_initialized_notification"
		return successResponse(request.ID, result)
	case "ping", "tools/list", "resources/list", "resources/read", "tools/call":
		if *phase != "ready" {
			return failureResponse(request.ID, -32002, "MCP session is not ready; complete initialization first.", nil)
		}
		return handleJSONRPCRequest(request, server, context)
	default:
		return handleJSONRPCRequest(request, server, context)
	}
}

func writeJSONRPCResponse(writer *bufio.Writer, response jsonRPCResponse) error {
	encoded, err := json.Marshal(response)
	if err != nil {
		return err
	}

	if _, err := writer.WriteString(string(encoded) + "\n"); err != nil {
		return err
	}

	return writer.Flush()
}

File: examples/python/apps/agent/mcp_stdio.py
# ==========================================================================
# APP v1.1.1
# apps/agent/mcp_stdio.py
# --------------------------------------------------------------------------
# MCP stdio transport adapter.
#
# Reads JSON-RPC messages from stdin, dispatches to AppMcpServer,
# and writes responses to stdout.
#
# Session phases: awaiting_initialize -> awaiting_initialized -> ready
# ==========================================================================

from __future__ import annotations

import json
import sys
from typing import Any

from core.shared.app_mcp_contracts import (
    AppMcpProtocolError,
    AppMcpRequestContext,
    AppMcpServer,
    BaseAppMcpProcessAdapter,
)

_JSONRPC_VERSION = "2.0"
_PARSE_ERROR = -32700
_INVALID_REQUEST = -32600
_METHOD_NOT_FOUND = -32601
_INTERNAL_ERROR = -32603


class StdioAppMcpAdapter(BaseAppMcpProcessAdapter):
    """MCP adapter for stdio (stdin/stdout) transport."""

    @property
    def transport(self) -> str:
        return "stdio"

    def serve(self, server: AppMcpServer) -> None:
        """
        Start serving MCP requests from stdin.

        Reads one JSON-RPC message per line, dispatches to the server,
        and writes responses to stdout. Exits when stdin closes.
        """
        phase = "awaiting_initialize"

        for line in sys.stdin:
            line = line.strip()
            if not line:
                continue

            # Parse JSON
            try:
                msg = json.loads(line)
            except json.JSONDecodeError:
                _write_error(None, _PARSE_ERROR, "Parse error")
                continue

            if not isinstance(msg, dict):
                _write_error(None, _INVALID_REQUEST, "Request must be a JSON object")
                continue

            # Reject batches
            if isinstance(msg, list):
                _write_error(None, _INVALID_REQUEST, "Batch requests not supported")
                continue

            method = msg.get("method")
            request_id = msg.get("id")
            params = msg.get("params", {})
            is_notification = request_id is None

            ctx = AppMcpRequestContext(
                transport="stdio",
                request_id=request_id,
            )

            # -- Phase: awaiting_initialize --
            if phase == "awaiting_initialize":
                if method == "initialize":
                    try:
                        result = server.initialize(params, ctx)
                        _write_result(request_id, result)
                        phase = "awaiting_initialized"
                    except AppMcpProtocolError as e:
                        _write_error(request_id, e.code, str(e), e.data)
                    continue

                if is_notification:
                    continue  # Silently drop notifications before init.

                _write_error(
                    request_id, _INVALID_REQUEST,
                    "Server not initialized. Send initialize first.",
                )
                continue

            # -- Phase: awaiting_initialized --
            if phase == "awaiting_initialized":
                if method == "notifications/initialized":
                    phase = "ready"
                    continue
                # Allow immediate requests (lenient, like the TS example).
                phase = "ready"

            # -- Phase: ready --
            if is_notification:
                # Handle known notifications silently.
                if method in ("notifications/initialized", "notifications/cancelled"):
                    continue
                continue

            # Dispatch request
            try:
                if method == "ping":
                    _write_result(request_id, {})

                elif method == "tools/list":
                    tools = server.list_tools(ctx)
                    _write_result(request_id, {"tools": tools})

                elif method == "resources/list":
                    resources = server.list_resources(ctx)
                    _write_result(request_id, {"resources": resources})

                elif method == "resources/read":
                    uri = (params or {}).get("uri", "")
                    result = server.read_resource(uri, ctx)
                    _write_result(request_id, result)

                elif method == "tools/call":
                    name = (params or {}).get("name", "")
                    args = (params or {}).get("arguments", {})
                    result = server.call_tool(name, args, ctx)
                    _write_result(request_id, result)

                elif method == "initialize":
                    _write_error(
                        request_id, _INVALID_REQUEST,
                        "Already initialized",
                    )

                else:
                    _write_error(
                        request_id, _METHOD_NOT_FOUND,
                        f"Unknown method: {method}",
                    )

            except AppMcpProtocolError as e:
                _write_error(request_id, e.code, str(e), e.data)
            except Exception as e:
                _write_error(request_id, _INTERNAL_ERROR, str(e))


def _write_result(request_id: Any, result: Any) -> None:
    """Write a JSON-RPC success response to stdout."""
    msg = {
        "jsonrpc": _JSONRPC_VERSION,
        "id": request_id,
        "result": result,
    }
    sys.stdout.write(json.dumps(msg, ensure_ascii=False) + "\n")
    sys.stdout.flush()


def _write_error(
    request_id: Any,
    code: int,
    message: str,
    data: Any = None,
) -> None:
    """Write a JSON-RPC error response to stdout."""
    error: dict[str, Any] = {"code": code, "message": message}
    if data is not None:
        error["data"] = data
    msg: dict[str, Any] = {
        "jsonrpc": _JSONRPC_VERSION,
        "id": request_id,
        "error": error,
    }
    sys.stdout.write(json.dumps(msg, ensure_ascii=False) + "\n")
    sys.stdout.flush()
